### ISSUE TITLE
Add fuzzy finder for bookmarks

### ISSUE THREAD

**Issue by D1mon** [philc/vimium#2457]
When you press "o" opens the input search. Usual search is used. It would be nice to work fuzzy finder search.

**Comment by innaterebel** [comment]
<kbd>b</kbd>／<kbd>B</kbd>?

**Comment by D1mon** [comment]
b - Open a bookmark
B - Open a bookmark in a new tab
o - Open URL, bookmark or history entry
O - Open URL, bookmark or history entry in a new tab

**Comment by binarious** [comment]
Dupe of https://github.com/philc/vimium/issues/2087. Would be a great feature.

@smblott-github, @mrmr1993  I'm not familiar with the code. Would https://github.com/philc/vimium/blob/master/content_scripts/mode_find.coffee be the place to implement this?

**Comment by mrmr1993** [comment]
@binarious You'll probably want to change the code that provides the results, which will be in one of the files in `background_scripts` with `competion` in its name (most likely [`completion.coffee`](https://github.com/philc/vimium/blob/master/background_scripts))./completion.coffee

**Comment by smblott-github** [comment]
You would start here: https://github.com/philc/vimium/blob/master/background_scripts/completion.coffee#L647.

But there would be knock-on consequences in places such as the code that highlights the matched text.

We probably would not want fuzzy matches displacing actual matches, particularly for bookmarks.

**Comment by toraritte** [comment]
Probably duplicate of #2450 

@smblott-github 
* How about using special characters to control the search? Something along the lines of https://github.com/junegunn/fzf#search-syntax with the exception that Vimium users expect the literal search to be default
* Would it be possible to create a split column below the search bar where one would show the literal results and the other the fuzzy matches?

**Comment by smblott-github** [comment]
> How about using special characters to control the search?

No.  The user shouldn't have to learn anything, it should just work.

> one would show the literal results and the other the fuzzy matches

That's not how I would do it.  I'd make sure that the scores for fuzzy matches are always lower than those for proper matches, then I'd just throw them in with everything else.

**Comment by MIvanchev** [comment]
This would be an awesome feature to have, how come it's not being discussed more actively :D

**Comment by DusanLesan** [comment]
This is the feature I miss the most. I know there is a separate issue for searching tabs, but I think all of the search fields should support fuzzy search. If the concern is that some users might not like it, an option could be added under a new miscellaneous entry to control it.

**Comment by forestchen1224** [comment]
There is a js implementation of fzf [https://github.com/ajitid/fzf-for-js](url)

**Comment by philc** [comment]
I know this is has been requested in several places over the years and each of us loves our completion algorithms.

My experience: I abandoned using fzf-like fuzzy matching (where no spaces are required between search terms) in my coding workflows years ago, because I found it too noisy. I think substring search with spaces in between each term works well: it's both high recall and high precision, i.e. it generally surfaces what you were looking for, without much additional noise. And it's unsurprising to users.
